apps/src/blockly/addons/cdoUtils.js

```diff
@@ -106,9 +109,168 @@ export function getUserTheme(themeOption) {
+  Blockly.serialization.workspaces.load(blockJson, tempJsonWorkspace);
+  const jsonBlocks = tempJsonWorkspace.getAllBlocks();
+
+  // Create an array of blocks based on the XML encoding of the current workspace.
+  const blockXml = Blockly.Xml.blockSpaceToDom(workspace);
+  const tempXmlWorkspace = new Blockly.Workspace();
+  Blockly.Xml.domToWorkspace(blockXml, tempXmlWorkspace);
+  const xmlBlocks = tempXmlWorkspace.getAllBlocks();
+
+  // compareBlockArrays returns an array of differences found.
+  const differences = compareBlockArrays(xmlBlocks, jsonBlocks);
+  const result =
+    differences.length > 0 ? FIREHOSE_EVENT_FAIL : FIREHOSE_EVENT_PASS;
+  // Log a record to Firehose/Redshift.
+  const recordData = {
+    projectUrl: dashboard.project.getShareUrl(),
+    blocklyVersion: Blockly.blockly_.VERSION,
+    differences: differences,
+  };
+  firehoseClient.putRecord(
+    {
+      study: FIREHOSE_STUDY,
+      event: result,
+      data_json: JSON.stringify(recordData),
+    },
+    {
+      alwaysPut: true, // Allows logging from development environments
+      includeUserId: true,
+    }
+  );
+  if (experiments.isEnabled(experiments.BLOCKLY_JSON)) {
+    // Avoid logging to the console for typical users.
+    console.log(recordData);
+  }
+}
+
+// Used to find differences between blocks created from xml and json sources.
+// Called when the app is initialized and each time the app runs.
+// Compares two block arrays and returns an array of differences between them.
+export function compareBlockArrays(xmlBlocks, jsonBlocks) {
+  // compareValues() will be called recursively so we need to keep track of objects
+  // that it has already compared, to prevent infinite loops caused by circular
+  // references.
+  const visited = new Set();
+  // Some block properties are expected to hold different values, or refer to blocks already in the array.
+  const keysToSkip = [
+    'parentBlock_', // a circular reference to another block in the array
+    'sourceBlock_', // a circular reference to another block in the array
+    'workspace', // a reference to the unique temporary workspaces used to create the two block arrays
+  ];
+
+  function compareValues(arg1, arg2, path = '', differences = []) {
+    if (arg1 === arg2) {
+      return differences;
+    }
+
+    if (arg1 && arg2 && typeof arg1 === 'object' && typeof arg2 === 'object') {
+      if (visited.has(arg1) || visited.has(arg2)) {
+        return differences; // Circular reference found, skip this object
+      }
+      visited.add(arg1);
+      visited.add(arg2);
+
+      const keys1 = Object.keys(arg1).sort();
+      const keys2 = Object.keys(arg2).sort();
+
+      if (keys1.length !== keys2.length) {
+        differences.push({
+          result: 'different length',
+          path: path,
+          key1: keys1.length,
+          key2: keys2.length,
+        });
+      }
+
+      for (let i = 0; i < keys1.length; i++) {
+        const key1 = keys1[i];
```

Comment:
In the array returned from the call to `getAllBlocks`, are the blocks always returned in the same sequence? I'm thinking of detached/unused blocks or programs with 2 or more event blocks.

Reply:
They should be, and I've tested with workspaces with multiple top-level blocks.  If they were not in the same order for any reason, the utility will return a difference.